### ISSUE TITLE
[BUG][Fuzzer][wrong-code] Sparse WGMMA `T.gemm_sp` with `block_N ≡ 8 (mod 16)` silently miscomputes the whole output tile on sm_90 (and hangs under warp specialization)

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2603]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11

### System information

- Installed: pip (dev build reporting `0.1.11`)
- GPU: NVIDIA H100 (sm_90a)
- Python 3.12, `torch 2.12.1+cu130`, CUDA 13

### Problem description

On sm_90, a 2:4-structured-sparse WGMMA `T.gemm_sp` whose N tile is congruent to 8 mod 16 (`block_N ∈ {24, 40, 56, 72, …}`) **compiles clean, runs with no error or warning, and returns a silently wrong result** — about 96–97% of the output-tile elements are incorrect. There is no exception and no diagnostic, and the result is deterministic (bit-identical across two runs at the same seed, and reproduced at independent seeds), so this is a compile/layout miscompute, not a race.

The trigger is `block_N ≡ 8 (mod 16)` **with more than one N column-atom** on the sparse-WGMMA path with warp specialization disabled (`pass_configs={"tl.disable_warp_specialized": True}`). `block_N` a multiple of 16 is correct; a single `block_N = 8` (one `n=8` sparse atom, `num_inst_n = 1`) is also correct.

The same `block_N ≡ 8 (mod 16)` config additionally **hangs at launch** under the default (warp-specialized) pipeline — the kernel launches, the GPU pins at 100%, and it never returns (see the WS-on row below).

A few things suggest `block_N ≡ 8 (mod 16)` is a shape this path is meant to handle rather than a misuse: the admission guard accepts any `N % 8 == 0` (`ComputeWgmmaWarpPartition`, `kNPerWarp = 8`), the sparse macro generator selects an `inst_n = gcd(warp_col_tiles, 256) = 8` atom for it, Hopper sparse WGMMA has native `m64n{24,40,56,72}k32` atoms, and the single-atom `block_N = 8` case already computes bit-exact — so the shape itself is computable and the wrong result is confined to the multi-column `n_dim = 8` unswizzled descriptor. (By contrast the M axis is guarded at `% 16` and `block_M = 8` is cleanly rejected; the N axis is guarded only at `% 8`.)

Reproduced on H100 (sm_90a). This is the sparse-WGMMA path only; sm_89 and earlier cannot reach this code.

<details>
<summary>N sweep, controls, the WS-on hang, and the emitted sparse-WGMMA atom (v0.1.11, H100, fp16 A / int16 metadata / fp32 accum, integer-valued inputs so exact equality is a valid check, each cell run x2)</summary>

Kernel: `M=64, K=128, block_M=64, block_K=64, threads=128`, one N-block (`block_N=N`), `Square` policy, `num_stages=2`, `transpose_B=False` (MN-major B).

WS **disabled** (`tl.disable_warp_specialized: True`):

```
N%16 | N   | mismatches / total   | emitted atom      | verdict
  0  |  16 |     0 / 1024         | n=16              | CORRECT (bit-exact)
  0  |  32 |     0 / 2048         | n=32              | CORRECT (bit-exact)
  0  |  64 |     0 / 4096         | n=64              | CORRECT (bit-exact)
  0  | 128 |     0 / 8192         | n=128             | CORRECT (bit-exact)
  8  |   8 |     0 / 512          | n=8, num_inst_n=1 | CORRECT (single atom)
  8  |  24 |  1491 / 1536         | n=8, num_inst_n=3 | WRONG (det=True)
  8  |  40 |  2491 / 2560         | n=8, num_inst_n=5 | WRONG (det=True)
  8  |  56 |  3478 / 3584         | n=8, num_inst_n=7 | WRONG (det=True)
  8  |  72 |  4450 / 4608         | n=8, num_inst_n=9 | WRONG (det=True)
```

Determinism: `N=24` gives 1491/1536 at seed 1234 and 1490/1536 at seed 42; `N=72` gives 4450/4608 at seed 1234 and 4485/4608 at seed 9999 — bit-identical across the two runs at each seed. `num_stages=0` and `num_stages=2` give the identical wrong result (4450/4608 for N=72), so this is stage-independent (a layout/descriptor error, not a pipeline-versioning one). The wrong rows are whole-row wrong, not a tail-drop — e.g. `N=24` row 0: `ref=[96,89,92,77,97,98,105,93]` vs `got=[99,103,95,103,95,74,73,92]`.

Controls (bit-exact, `nmis=0`, det): every `N % 16 == 0` (16/32/64/128) and `N = 8` (single `n=8` atom). So the trigger is specifically `n_dim = 8` **with `num_inst_n > 1`**.

A plain `global → shared → global` `T.copy` of a `64×24` and a `64×72` tile is EXACT (`0/3072`, `0/9216`) over the same inputs — so the fault is in the `T.gemm_sp` compute/layout, not in copying B into shared memory.

WS **enabled** (default pipeline, `num_stages=2`), same kernel, exit code of a 75 s-bounded run:

```
N   | exit | result
 64 |   0  | mism=0/4096, no hang (CORRECT)
128 |   0  | mism=0/8192, no hang (CORRECT)
  8 | 124  | HANG (no result)
 24 | 124  | HANG (no result)
 40 | 124  | HANG (no result)
 56 | 124  | HANG (no result)
 72 | 124  | HANG (no result)
```

`N=8` is EXACT under WS-off but HANGS under WS-on, so the deadlock is a distinct face from the compute miscompute — it is a warp-specialized TMA-arrive deadlock on the small, non-swizzleable B tile (the `block_N ≡ 8 (mod 16)` B shared tile has an inner width that can't take a clean swizzle and is routed to the 1-D bulk TMA copy).

Emitted CUDA (the wrong-vs-right atom is directly visible in the kernel source):
```
N=24 (block_N ≡ 8 mod 16): tl::wgmma_sp_ss<..., 64, 8, 32, ...>    # n_dim=8 atom, num_inst_n=3
N=64 (control):            tl::wgmma_sp_ss<..., 64, 64, 32, ...>   # n_dim=64, correct
```
</details>

### Reproducible example code

```python
import tilelang, tilelang.language as T, torch
from tilelang.utils.sparse import compress, get_e_factor

E_FACTOR = get_e_factor(T.float16, T.int16)

def make_sp_gemm(M, N, K, bM, bN, bK):
    @T.prim_func
    def main(
        A_sparse: T.Tensor((M, K // 2), "float16"),      # 2:4-compressed A
        E: T.Tensor((M, K // E_FACTOR), "int16"),        # 2:4 metadata
        B: T.Tensor((K, N), "float16"),                  # dense, MN-major
        C: T.Tensor((M, N), "float32"),
    ):
        with T.Kernel(T.ceildiv(N, bN), T.ceildiv(M, bM), threads=128) as (bx, by):
            As = T.alloc_shared((bM, bK // 2), "float16")
            Es = T.alloc_shared((bM, bK // E_FACTOR), "int16")
            Bs = T.alloc_shared((bK, bN), "float16")
            Cf = T.alloc_fragment((bM, bN), "float32")
            T.clear(Cf)
            for ko in T.Pipelined(T.ceildiv(K, bK), num_stages=2):
                T.copy(E[by * bM, ko * bK // E_FACTOR], Es)
                T.copy(A_sparse[by * bM, ko * bK // 2], As)
                T.copy(B[ko * bK, bx * bN], Bs)
                T.gemm_sp(As, Es, Bs, Cf, transpose_A=False, transpose_E=False,
                          transpose_B=False, policy=T.GemmWarpPolicy.Square)
            T.copy(Cf, C[by * bM, bx * bN])
    return main

def run(N):
    M, K = 64, 128
    g = torch.Generator(device="cuda").manual_seed(1234)
    # exact 2:4 pattern (keep 2 of every 4 K-cols via randn topk, no int ties),
    # then set the kept values to ints in {1,2} so fp32 accum is exact.
    r = torch.randn((M, K // 4, 4), generator=g, device="cuda")
    mask = torch.zeros_like(r, dtype=torch.bool)
    mask.scatter_(-1, r.topk(2, dim=-1).indices, True)
    vals = torch.randint(1, 3, (M, K // 4, 4), generator=g, device="cuda").float()
    A = (vals * mask).view(M, K).half()
    B = torch.randint(0, 3, (K, N), generator=g, device="cuda").float().half()

    A_sparse, Emeta = compress(A.contiguous(), meta_dtype=torch.int16)
    kernel = tilelang.compile(
        make_sp_gemm(M, N, K, 64, N, 64), out_idx=[3],
        pass_configs={"tl.disable_warp_specialized": True})
    out = kernel(A_sparse, Emeta, B).float()
    ref = A.float() @ B.float()                 # dense reference over densified A
    return (out != ref).sum().item(), out.numel()

print("N=64 (mult of 16):", run(64))    # -> (0, 4096)     CORRECT
print("N=24 (8 mod 16): ", run(24))     # -> (1491, 1536)  WRONG, silent
```

### Traceback

```pytb
No traceback — with warp specialization disabled the kernel compiles and runs to completion; the result is silently wrong and deterministic, with no error or warning. Under the default (warp-specialized) pipeline the same config hangs at launch and never returns (no diagnostic).
```

### Expected behavior

A `block_N ≡ 8 (mod 16)` sparse GEMM shouldn't silently return a wrong tile (or hang). The shape appears to be in scope — it's admitted by the `N % 8` guard, selects a dedicated `n = 8` sparse atom, maps to native sparse WGMMA `m64n{24,…}k32` atoms, and the single-atom `block_N = 8` case already computes bit-exact — so computing the correct result for the multi-column case as well seems the natural expectation. If that descriptor is hard to fix, a clean compile-time rejection would at least be preferable to a silent wrong result / hang. Either way, running to completion and returning silently-wrong numbers (or deadlocking with no diagnostic) is the problem.

### Additional context

**Root cause.** The wrong result is a B shared-memory descriptor field that fails to scale with N on the unswizzled (`SWIZZLE_NONE`) layout of the sparse macro generator — structurally the same shape of bug as on the dense path, but in the separately-authored sparse file. Two things line up for `block_N ≡ 8 (mod 16)` on fp16:

1. **A dedicated `n = 8` atom is selected and tiled `num_inst_n > 1` times.** [`_initialize_wgmma_prefix`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/tilelang/cuda/intrinsics/macro/wgmma_sp_macro_generator.py#L89-L102) sets `inst_n = gcd(warp_col_tiles, 256)`, which is `8` when `warp_col_tiles ≡ 8 (mod 16)`; the emit then tiles `num_inst_n = warp_col_tiles // 8` such atoms along N (e.g. 3 atoms for N=24).

2. **The `SWIZZLE_NONE` B-descriptor puts the N-scaled value in the wrong field.** For the MN-major (`b_is_k_major = False`) unswizzled case, [`compute_wgmma_b_desc_params`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/tilelang/cuda/intrinsics/macro/wgmma_sp_macro_generator.py#L183-L211) emits `b_leading_byte_offset = 8 * n_dim * elems_in_bytes` (the correctly N-scaled `8·N`, but deposited in **LBO**) and `b_stride_byte_offset = (0 if n_dim == 8 else 8 * 8 * elems_in_bytes)` — a **constant, N-independent** value in **SBO**. The SBO for a row-major `(K,N)` tile is the stride from the first 8 K-rows to the next 8 K-rows, which should scale with N; hardwiring it to `8*8*elems` is correct only at `n_dim == 8` (the single-atom case, where SBO is `0`). For every wider `N ≡ 8 (mod 16)` the multi-atom emit walks the wrong K-row pitch and each WGMMA reads the wrong shared-memory bytes — a uniform, whole-tile corruption (each output is a well-formed dot over the wrong set of B elements, not a partial sum or a dropped column band). Verified on H100: the emitted descriptor takes the `SWIZZLE_NONE` branch for these N, and the single-atom `block_N = 8` is bit-exact while `block_N = 24/40/56/72` is not.

The warp-partition helper [`ComputeWgmmaWarpPartition`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/op/gemm_sp.cc#L182-L194) guards only [`N % 8 == 0`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/op/gemm_sp.cc#L188) (M uses `% 16`), so `N ≡ 8 (mod 16)` is admitted with no reject.

The **WS-on hang** is a separate face on the same trigger: the `block_N ≡ 8 (mod 16)` B shared tile is not cleanly swizzleable, so under warp specialization its TMA load is routed to the 1-D bulk copy path; the forward-barrier arrival is not emitted on that route and the consumer wait never flips phase, so the kernel deadlocks (this matches the mbarrier-arrive-drop mechanism, reached here via the sparse path and the `block_N` trigger rather than a single-iteration loop).

<details>
<summary>Possible fixes (proposed, not built) and disposition</summary>

Since `block_N ≡ 8 (mod 16)` is admitted by the `N % 8` guard and the single-atom `block_N = 8` already computes correctly, one direction is to make the multi-column MN-major unswizzled path compute correctly too: the `SWIZZLE_NONE` B-descriptor's SBO would need to scale its K-row pitch with N (reconciling the LBO/SBO roles against the PTX MN-major core-matrix definition). Alternatively, if that path is not intended to carry these shapes, tightening the `N % 8` guard in `ComputeWgmmaWarpPartition` (`gemm_sp.cc`) to `N % 16` would turn the silent wrong answer (and the hang) into a clean compile-time error. Which is appropriate depends on whether the multi-column `n = 8` MN-major sparse path is meant to support `N ≡ 8 (mod 16)` — a call for the maintainers.
</details>

**Provenance.** The sparse WGMMA macro generator, including the `b_stride_byte_offset = (0 if n_dim == 8 else 8*8*elems)` descriptor line, was added in [#2048](https://github.com/tile-ai/tilelang/pull/2048) ("[Backend] Refactor gemm_sp", merged 2026-05-19); `ComputeWgmmaWarpPartition`'s `N % 8` guard moved into `src/cuda/op/gemm_sp.cc` in [#2297](https://github.com/tile-ai/tilelang/pull/2297) (a namespace refactor). Not a regression — the multi-column `n = 8` sparse descriptor has been wrong since the refactored sparse WGMMA path first shipped. The project's own sparse-GEMM tests and the `examples/gemm_sp` example use N tiles that are multiples of 16 (or the single `N = 8`), so the `≡ 8 (mod 16)` multi-atom case is never exercised.

**Related.** The dense `T.gemm` path has the analogous N-axis `n_dim = 8` miscompute filed as [#2593](https://github.com/tile-ai/tilelang/issues/2593) (in the separate `wgmma_macro_generator.py`); this report is the sparse `T.gemm_sp` path, a distinct file and lowering (`src/cuda/op/gemm_sp.cc`, `tilelang/cuda/intrinsics/macro/wgmma_sp_macro_generator.py`, `tilelang/tileop/gemm_sp/`). It is distinct from the dense N-warp-partition band-drop [#2537](https://github.com/tile-ai/tilelang/issues/2537) (there `warp_col_tiles = N // n_warp` floors and leaves a trailing column band unwritten; here `n_warp = 1`, `warp_col_tiles = N` exactly, no floor loss, and the whole tile is corrupted — a layout error, not a coverage gap), from the dense M-axis row-swap [#2522](https://github.com/tile-ai/tilelang/issues/2522) (that is the M axis; this is the N axis), and from the single-iteration WS deadlock [#2566](https://github.com/tile-ai/tilelang/issues/2566) (that fires at trip count 1; the WS-on hang here fires on `block_N ≡ 8 (mod 16)` at any trip count). I searched the open and closed tracker and found no existing report of this silent miscompute (or hang) on the sparse `T.gemm_sp` N-axis `n = 8` path.
